components/ai_chat/ai_chat_api.cc

```diff
@@ -113,40 +149,78 @@ void AIChatAPI::QueryPrompt(ResponseCallback callback,
   dict.Set("top_p", 0.999);
   dict.Set("model", model_name);
   dict.Set("stop_sequences", std::move(stop_sequences));
-  dict.Set("stream", false);
+  dict.Set("stream", true);
 
-  base::flat_map<std::string, std::string> headers;
-  headers.emplace("x-brave-key", BUILDFLAG(BRAVE_SERVICES_KEY));
+  return dict;
+}
+
+void AIChatAPI::OnDataReceived(base::StringPiece string_piece,
+                               base::OnceClosure resume) {
+  std::vector<std::string> stream_data = base::SplitString(
+      string_piece, "\r\n", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+  static constexpr char kDataPrefix[] = "data: {";
 
-  DVLOG(1) << __func__ << " Prompt: " << prompt << "\n";
-  DVLOG(1) << __func__ << " Using model: " << model_name;
+  // Binding a remote will guarantee the order of the messages being sent, so we
+  // must ensure that it is connected.
+  DCHECK(data_decoder_);
+  DCHECK(json_parser_.is_connected());
 
-  api_request_helper_.Request("POST", api_url, CreateJSONRequestBody(dict),
-                              "application/json", std::move(internal_callback),
-                              headers);
+  for (const auto& data : stream_data) {
+    if (!base::StartsWith(data, kDataPrefix)) {
+      continue;
+    }
+
+    std::string json = data.substr(strlen(kDataPrefix) - 1);
```

Comment:
I'm curious if there isn't closing bracket at the end of |data|. And it might be good to have a comment about what form is expected for |data|, such as

```
// Example of |data|'s contents
// 
// "data: { foo: bar }"
```

Reply:
There is a closing bracket, however we assume the format from `data: {...` onwards is going to be json.

Reply:
Thanks, I was mistaken about the output of `data.substr(...)`. I missed the '-1' and thought the output doesn't start with opening bracket 🤪